### PR TITLE
[ModelRunner V2] Simplify penalties implementation

### PR DESCRIPTION
**Pull request by njhill**
Avoid separate `expanded_local_pos` input batch state.

cc @izhuhaoran

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/penalties.py [hunk 1]
```diff
@@ -141,8 +138,15 @@ def _penalties_kernel(
         other=0,
     )
 
-    # Compute cumulative draft_counts from previous positions in this request
-    pos = tl.load(expanded_local_pos_ptr + token_idx)
+    # Derive local position within this request's logit group by scanning
+    # backwards in expanded_idx_mapping. Since req_state_idx is unique per
+    # active request, a change in value marks the request boundary.
```
Comment (izhuhaoran) [inline vllm/v1/worker/gpu/sample/penalties.py:143]: I actually considered this approach when implementing the kernel, but was worried it might regress performance. Could you do a benchmark comparing the latency against the original `expanded_local_pos_ptr` approach under large batch_size / num_tokens?

File: vllm/v1/worker/gpu/sample/penalties.py [hunk 2]
```diff
@@ -141,8 +138,15 @@ def _penalties_kernel(
...
+    pos = 0
+    for i in tl.static_range(MAX_SPEC_LEN):
+        check_idx = token_idx - i - 1
+        if check_idx >= 0:
+            prev_req = tl.load(idx_mapping_ptr + check_idx)
+            pos += (prev_req == req_state_idx).to(tl.int32)
```
Comment (izhuhaoran) [inline vllm/v1/worker/gpu/sample/penalties.py]: I'm not sure whether this would be better ?
```suggestion
        prev_req = tl.load(idx_mapping_ptr + check_idx, mask=check_idx >= 0, other=-1)
        pos += (prev_req == req_state_idx).to(tl.int32)
```

**Review by izhuhaoran — COMMENTED** [review]
@njhill Thanks for this nice job, some small comments

File: vllm/v1/worker/gpu/sample/penalties.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/sample/penalties.py]: +1, same concern here.

File: vllm/v1/worker/gpu/sample/penalties.py (hunk 1 above)
Comment (izhuhaoran) [reply vllm/v1/worker/gpu/sample/penalties.py:143]: BTW, `expanded_local_pos` is also needed by bad_words (see https://github.com/vllm-project/vllm/pull/33433). Removing it here would lead to duplicated computation across penalty and badwords kernels.

Comment (njhill) [reply vllm/v1/worker/gpu/sample/penalties.py:143]: Thanks @izhuhaoran. Agree we should benchmark this, though since penalties aren't frequently used, if the perf difference is negligible it might still be worth the simplification. Bad words is even less frequently used AFAIK.

But I'm also thinking about pluggable logits processors - the field may be beneficial for those to support spec decoding more easily.

So I'm not saying we should definitely merge this change. There's no rush, we can think about more and test it some more.

Comment (izhuhaoran) [reply vllm/v1/worker/gpu/sample/penalties.py:143]: > So I'm not saying we should definitely merge this change. There's no rush, we can think about more and test it some more.

Agree with this.
